Don't access system indices directly in EsRestTestCase

williamrandolph opened the pull request:
We have a wildcard deletion in out `EsRestTestCase` class that can trigger system index access deprecation warnings. If we invoke the feature reset API before that deletion, we should be able to avoid the warnings.

Relates https://github.com/elastic/elasticsearch/issues/62501

williamrandolph committed:
Use feature reset to avoid system index warnings

williamrandolph committed:
Try shielding against mixed clusters

williamrandolph committed:
Avoid extensive ML cleanup when ML is disabled

williamrandolph committed:
Merge branch 'master' into es-rest-test-system-indices

elasticmachine committed:
Merge branch 'master' into es-rest-test-system-indices

elasticmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

williamrandolph commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/MachineLearning.java:
```diff
@@ -1298,6 +1298,11 @@ public void cleanUpFeature(
         ActionListener<ResetFeatureStateResponse.ResetFeatureStateStatus> finalListener) {
         logger.info("Starting machine learning feature reset");
 
+        if (enabled == false) {
```
@droberts195 Could you or someone else from the Machine Learning team let me know what you think of this? The problem I ran into was that when the ML plugin is disabled, `SetResetActionModeRequest` and other such actions aren't provided by the plugin, so the client fails to execute. It seems to me that if ML is disabled, we can just skip everything below, but I'd like to be sure about that.

droberts195 replied:
Yes, I think you are correct.  It's a requirement that if ML is being used then the plugin must be enabled on all master-eligible nodes.  If this rule has been obeyed then all the code below would be no-ops.  And if the rule hasn't been obeyed then the cluster is into undefined behaviour territory and other things will likely be going wrong anyway.

I agree it's still work calling the super class method because somebody might have used ML, then stopped using and deleted all the jobs and datafeeds before disabling the plugin, yet still have a few indices left over, so it's nice to clean those up.

gwbrown commented on test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:
```diff
@@ -748,25 +748,38 @@ private static void deleteAllNodeShutdownMetadata() throws IOException {
-                }).build();
-            deleteRequest.setOptions(allowSystemIndexAccessWarningOptions);
+
+            if (minimumNodeVersion().onOrAfter(Version.CURRENT)) {
```
`Version.CURRENT` will change when we cut a new release, so this branch will only ever be taken in non-mixed clusters. Shouldn't we check against the version the feature reset API was implemented? (or whatever version it's functional enough to be usable in at least)

williamrandolph replied:
That's totally correct. I used `Version.CURRENT` in order to delay thinking about BWC issues and merging, but I didn't even put a TODO in, so it would have been easy to forget.

- [ ] put in a definite version, or a clear BWC todo comment

gwbrown commented on test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:
```diff
@@ -748,25 +748,38 @@ private static void deleteAllNodeShutdownMetadata() throws IOException {
+            if (minimumNodeVersion().onOrAfter(Version.CURRENT)) {
+                // feature reset deletes system indices
+                final Request postRequest = new Request("POST", "/_features/_reset");
+                adminClient().performRequest(postRequest);
```
We might want to check the response, since it's possible for this request to only partially succeed. That shouldn't happen in a test, but if it does, I'd rather get a stack trace pointing here than have another test fail with a really unclear error.

droberts195 replied:
+100 on logging both exceptions and partial failures here

> That shouldn't happen in a test

But it will, especially in PR builds, because somebody will make a false assumption about how their code changes interact with feature reset.  The feature reset API will become one of the most heavily tested APIs in the product when this PR is merged.  All the other REST tests that exist will exercise an enormous number of edge cases in feature reset (which is good, but we should plan for failures).

williamrandolph replied:
I've updated this so that we throw an error when feature reset fails. I expected that the HTTP Entity would be the full response message we return on an error, but it's just a root cause message. But at least now we have a message that indicates a problem with resetting features.

gwbrown commented on test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:
```diff
@@ -748,25 +748,38 @@ private static void deleteAllNodeShutdownMetadata() throws IOException {
+                final Request postRequest = new Request("POST", "/_features/_reset");
+                adminClient().performRequest(postRequest);
+            } else {
+                RequestOptions allowSystemIndexAccessWarningOptions = RequestOptions.DEFAULT.toBuilder()
+                    .setWarningsHandler(warnings -> {
+                        if (warnings.size() == 0) {
+                            return false;
+                        } else if (warnings.size() > 1) {
+                            return true;
+                        }
+                        // We don't know exactly which indices we're cleaning up in advance,
+                        // so just accept all system index access warnings.
+                        final String warning = warnings.get(0);
+                        final boolean isSystemIndexWarning = warning.contains("this request accesses system indices") &&
+                            warning.contains("but in a future major version, direct access to system indices will be prevented by default");
+                        return isSystemIndexWarning == false;
+                    }).build();
+                deleteRequest.setOptions(allowSystemIndexAccessWarningOptions);
```
Should we still swallow system index access warnings, given we're using the reset API prior to this request so there shouldn't be any system indices left to delete?

I think it's fine if you left this here on purpose to prevent surprise errors for now, but we should open an issue to clean it up later if that's the case.

williamrandolph replied:
This ought to be the code path where we have old nodes in the cluster that don't have a feature reset API. It will probably help to put in a specific version number and comment explaining that this only works when all of your nodes have the ability to reset features.

AthenaEryma reviewed (commented):
Thanks for picking this up, this is a good step forward. I left a few comments in places I have some concerns about, though.

droberts195 commented on test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:
```diff
@@ -748,25 +748,38 @@ private static void deleteAllNodeShutdownMetadata() throws IOException {
     }
 
     protected static void wipeAllIndices() throws IOException {
```
It's interesting that there is now a call to the feature reset API in a method that claims to wipe _indices_, because feature reset does more.

I am not suggesting that you tackle this in this PR, but maybe it's worth adding a TODO to revisit cleanup in between tests once we reach a version where feature reset is reliably available.

I think we have too many options for preserve this/preserve that between tests.  I suspect in reality only a small number of combinations of these preservation options are ever set, likely preserve everything (for upgrade tests) and preserve nothing (for almost all other tests).

Calling the feature reset API in this general method that always gets called highlights the inconsistencies.  For example, there's a flag for whether to preserve rollups between tests, but if the feature reset API removed rollups (as it logically should because it removes ML jobs) then calling the feature reset API would render that flag useless.

There are already inconsistencies in this area, because we already call the feature reset API for ML cleanup after ML tests.  There is a comment on that bit of code saying it will wipe more than just ML but it shouldn't matter because everything will be getting wiped between tests anyway.  Since that didn't cause any problems it shows that there's no test suite where both ML tests and non-ML tests run where the fine-grained options to preserve some things but not others are set.

After this PR is merged the cleanup sequence for ML tests will call the feature reset API twice, once to clean up ML, then later when wiping all indices.  That shouldn't break anything because the second call should be a no-op.

Anyway, like I said, don't change any code in this PR because of these issues, but it seems like it would be a good idea to add a TODO and a long term roadmap item to rationalise and simplify cleanup in between tests once feature reset is guaranteed to be available.

jaymode replied:
> it would be a good idea to add a TODO and a long term roadmap item to rationalise and simplify cleanup in between tests once feature reset is guaranteed to be available.

I suggest we do the above and make sure a GH issue is mentioned in the TODO for cleaning this up to help track the origin of the TODO

williamrandolph replied:
While cleaning this up for merge, it made sense to me to move the feature cleanup to the top level. I'll move it back if this causes any integration test problems.

williamrandolph committed:
Merge branch 'master' into es-rest-test-system-indices

williamrandolph committed:
Move feature reset out of reset indices method

williamrandolph committed:
Throw error on failed feature reset

williamrandolph committed:
Use wildcard for optional index deletion

williamrandolph committed:
Allow restricted indices in rest test roles

williamrandolph committed:
Allow restricted indices in security async tests

williamrandolph commented:
@elasticmachine please run elasticsearch-ci/part-2

williamrandolph committed:
Allow restricted indices in eql security test

williamrandolph committed:
Avoid resetting system indices in certain xpack tests

williamrandolph commented:
@elasticmachine test this please

williamrandolph committed:
Move feature state reset closer to index deletion

williamrandolph committed:
Skip feature state reset in multi cluster security test

williamrandolph committed:
Merge branch 'master' into es-rest-test-system-indices

williamrandolph committed:
Don't reset feature states for cluster restart bwc test

williamrandolph committed:
Preserve feature states in some bwc tests

williamrandolph committed:
Merge branch 'master' into es-rest-test-system-indices

williamrandolph committed:
Preserve feature states for rolling upgrade tests

williamrandolph committed:
Ignore warnings about watcher index

williamrandolph commented:
@elasticmachine test this please

williamrandolph commented:
@elasticmachine test this please

williamrandolph committed:
Merging master into es-rest-test-system-indices

williamrandolph commented:
@elasticmachine test this please

williamrandolph committed:
include colon in warning header

williamrandolph committed:
Merge branch 'master' into es-rest-test-system-indices

williamrandolph commented:
@elasticmachine test this please

williamrandolph commented:
@elasticmachine update this please

williamrandolph commented:
@elasticmachine update branch please

elasticmachine committed:
Merge branch 'master' into es-rest-test-system-indices

gwbrown commented on client/rest-high-level/roles.yml:
```diff
@@ -5,6 +5,7 @@ admin:
     - names: '*'
       privileges:
         - all
+      allow_restricted_indices: true
```
Is this so that we can delete `.watches` when it gets reinstated?

Similar question for the other instances of this - why do we need it? When we flip the switch to disallowing system index access via REST, can we remove these?

williamrandolph replied:
I did this because I was getting security errors on the feature reset API calls on tests where security is enabled. I will take another look at the feature reset API code and see if there's an approach where we can call the feature reset API without needing the caller to allow restricted indices.

At the very least I will add a test somewhere that validates the behavior I saw to make it easier to discuss options.

williamrandolph replied:
I've dug in on this and I don't really see a way around it. If security is enabled and you want to delete a restricted index, you need a role with `allow_restricted_indices: true`. So I'll just add the behavior-confirming test.

williamrandolph replied:
Since we are no longer allowing restricted indices in test roles, I'm going to resolve this conversation.

gwbrown commented on test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:
```diff
@@ -779,6 +808,38 @@ protected static void wipeAllIndices() throws IOException {
         }
     }
 
+    // The watcher index may be recreated between when we reset system features and when
+    // we call a wildcard deletion
```
Can we expand this comment with why we think this happens, and that this can be removed once `DELETE *` doesn't hit system indices anyway?

AthenaEryma reviewed (approved):
Left a question and a minor comment, but overall LGTM - no need for another round of review after those are addressed.

williamrandolph committed:
Add context to comment

williamrandolph committed:
Merge branch 'master' into es-rest-test-system-indices

williamrandolph committed:
Use system context for feature state reset

It was hard to answer questions about why we needed to allow restricted
indices in any user role that reset feature states. This commit uses a
different approach to the problem: we switch to a system thread context
before calling feature cleanup functions.

For this to work, we have to add more allowed actions to the system
role. If this is undesirable, perhaps we can look at how to use a
new or different role in feature state resetting.

Another approach we considered was wrapping the client in an
OriginSettingClient when calling a feature's cleanup action, but the
origins are stored at the system index descriptor level, not at the
plugin level, so this would be a trickier implementation.

williamrandolph committed:
Merge branch 'master' into es-rest-test-system-indices

williamrandolph committed:
Merge branch 'master' into es-rest-test-system-indices

williamrandolph committed:
Merge branch 'master' into es-rest-test-system-indices

williamrandolph committed:
Use wildcard for ML process kill permission

williamrandolph committed:
Merge branch 'master' into es-rest-test-system-indices

williamrandolph commented:
@elasticmachine test this please

williamrandolph committed:
Restrict feature reset API to superuser

williamrandolph commented on server/src/main/java/org/elasticsearch/action/admin/cluster/snapshots/features/TransportResetFeatureStateAction.java:
```diff
@@ -81,7 +82,12 @@ protected void masterOperation(
 
         for (SystemIndices.Feature feature : systemIndices.getFeatures().values()) {
-            feature.getCleanUpFunction().apply(clusterService, client, groupedActionListener);
+            // we have verified admin privileges, so issue deletions in system context, not user context
```
This is a great suggestion. I've updated the `MANAGE` privilege to exclude the feature reset API.

williamrandolph committed:
Merge branch 'master' into es-rest-test-system-indices

williamrandolph committed:
Improve API response when transform feature reset fails

williamrandolph committed:
Work around bug in transform reset

williamrandolph committed:
Merge branch 'master' into es-rest-test-system-indices

benwtrent commented on test/framework/src/main/java/org/elasticsearch/test/rest/ESRestTestCase.java:
```diff
@@ -745,26 +756,46 @@ protected void deleteAllNodeShutdownMetadata() throws IOException {
+            } catch (ResponseException e) {
+                response = e.getResponse();
+        }
+        if (response.getStatusLine().getStatusCode() != 200) {
+            // TODO: since there is a bug in transform reset (link to issue!) we can't throw an error here
+            // throw new IllegalStateException("Failed to reset all feature states: "
+            logger.warn("Failed to reset all feature states: "
+                + Streams.copyToString(new InputStreamReader(response.getEntity().getContent(), StandardCharsets.UTF_8)));
+            }
```
I am hoping this isn't the case any longer :). There were bugs at the start, but if there are some current ones, it would be good to know what they are.

williamrandolph replied:
I kept running into errors like this one (note that there are some investigation log lines in that output): https://gradle-enterprise.elastic.co/s/rdu2sojvv7njq/console-log?task=:x-pack:plugin:transform:qa:single-node-tests:javaRestTest

You should be able to reproduce it at commit e754de22529 with:
```
➜  ./gradlew ':x-pack:plugin:transform:qa:single-node-tests:javaRestTest' --tests "org.elasticsearch.xpack.transform.integration.TransformGetAndGetStatsIT.testGetStatsWithContinuous" -Dtests.locale=ro-RO -Dtests.timezone=Africa/Asmera -Druntime.java=11
```

I wasn't able to pin down exactly where the problem was, so I was going to take another stab at it before writing up an issue. At one point I did have a logger giving me the stack trace for the failure, for example here: https://gradle-enterprise.elastic.co/s/psveo5xaja3mc/console-log?task=:x-pack:plugin:transform:qa:single-node-tests:javaRestTest

Pasted below for your convenience...

<details><summary>big stack trace</summary>

```
» [2021-08-18T14:19:27,042][WARN ][r.suppressed             ] [javaRestTest-0] path: /_features/_reset, params: {}    
»  org.elasticsearch.ElasticsearchStatusException: Failed node [X0DmH_ZZSX-GFG-9I2ZcdQ]    
»      at org.elasticsearch.xpack.transform.action.TransportStopTransformAction.buildException(TransportStopTransformAction.java:342) ~[?:?]    
»      at org.elasticsearch.xpack.transform.action.TransportStopTransformAction.lambda$waitForStopListener$9(TransportStopTransformAction.java:317) ~[?:?]    
»      at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:134) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.support.tasks.TransportTasksAction$AsyncAction.finishHim(TransportTasksAction.java:300) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.support.tasks.TransportTasksAction$AsyncAction.onFailure(TransportTasksAction.java:287) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.support.tasks.TransportTasksAction$AsyncAction$1.handleException(TransportTasksAction.java:263) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.transport.TransportService$5.handleException(TransportService.java:629) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.transport.TransportService$ContextRestoreResponseHandler.handleException(TransportService.java:1174) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.transport.TransportService$DirectResponseChannel.processException(TransportService.java:1283) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:1257) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.transport.TaskTransportChannel.sendResponse(TaskTransportChannel.java:50) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler$1.onFailure(SecurityServerTransportInterceptor.java:198) ~[?:?]    
»      at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$AbstractFilterListener.onFailure(SecurityServerTransportInterceptor.java:292) ~[?:?]    
»      at org.elasticsearch.xpack.security.authz.AuthorizationService.authorizeSystemUser(AuthorizationService.java:452) ~[?:?]    
»      at org.elasticsearch.xpack.security.authz.AuthorizationService.authorize(AuthorizationService.java:226) ~[?:?]    
»      at org.elasticsearch.xpack.security.transport.ServerTransportFilter.lambda$inbound$1(ServerTransportFilter.java:109) ~[?:?]    
»      at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:134) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.xpack.security.authc.AuthenticationService$Authenticator.authenticateAsync(AuthenticationService.java:342) ~[?:?]    
»      at org.elasticsearch.xpack.security.authc.AuthenticationService.authenticate(AuthenticationService.java:184) ~[?:?]    
»      at org.elasticsearch.xpack.security.transport.ServerTransportFilter.inbound(ServerTransportFilter.java:100) ~[?:?]    
»      at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$ProfileSecuredRequestHandler.messageReceived(SecurityServerTransportInterceptor.java:275) ~[?:?]    
»      at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:61) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.transport.TransportService.sendLocalRequest(TransportService.java:792) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.transport.TransportService$3.sendRequest(TransportService.java:114) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.transport.TransportService.sendRequestInternal(TransportService.java:736) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor.sendWithUser(SecurityServerTransportInterceptor.java:126) ~[?:?]    
»      at org.elasticsearch.xpack.security.transport.SecurityServerTransportInterceptor$1.sendRequest(SecurityServerTransportInterceptor.java:109) ~[?:?]    
»      at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:650) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:576) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.support.tasks.TransportTasksAction$AsyncAction.start(TransportTasksAction.java:249) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.support.tasks.TransportTasksAction.doExecute(TransportTasksAction.java:83) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.xpack.transform.action.TransportStopTransformAction.lambda$doExecute$0(TransportStopTransformAction.java:188) ~[?:?]    
»      at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:134) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.xpack.transform.persistence.IndexBasedTransformConfigManager.lambda$expandTransformIds$10(IndexBasedTransformConfigManager.java:466) ~[?:?]    
»      at org.elasticsearch.action.ActionListener$1.onResponse(ActionListener.java:134) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:31) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.client.node.NodeClient.lambda$executeLocally$0(NodeClient.java:103) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:169) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.tasks.TaskManager$1.onResponse(TaskManager.java:163) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.support.ContextPreservingActionListener.onResponse(ContextPreservingActionListener.java:31) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.xpack.security.action.filter.SecurityActionFilter.lambda$applyInternal$2(SecurityActionFilter.java:146) ~[?:?]    
»      at org.elasticsearch.action.ActionListener$DelegatingFailureActionListener.onResponse(ActionListener.java:217) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.ActionListener$RunAfterActionListener.onResponse(ActionListener.java:339) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.search.AbstractSearchAsyncAction.sendSearchResponse(AbstractSearchAsyncAction.java:655) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.search.ExpandSearchPhase.run(ExpandSearchPhase.java:109) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.search.AbstractSearchAsyncAction.executePhase(AbstractSearchAsyncAction.java:424) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:418) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.search.FetchSearchPhase.moveToNextPhase(FetchSearchPhase.java:219) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.search.FetchSearchPhase.lambda$innerRun$1(FetchSearchPhase.java:101) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.search.FetchSearchPhase.innerRun(FetchSearchPhase.java:107) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.action.search.FetchSearchPhase$1.doRun(FetchSearchPhase.java:84) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:737) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) ~[elasticsearch-8.0.0-SNAPSHOT.jar:8.0.0-SNAPSHOT]    
»      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) [?:?]    
»      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) [?:?]    
»      at java.lang.Thread.run(Thread.java:829) [?:?]

```
</details>

williamrandolph committed:
Update docs/changelog/73734.yaml

elasticsearchmachine commented:
Hi @williamrandolph, I've created a changelog YAML for you.

dakrone commented:
@williamrandolph are you still planning to get back to this PR? I think it ends up causing a few test failures that affect our team, so it'd be great if we could rule out this cause of failures.

williamrandolph commented:
@dakrone I'll take a look at what it would take to get it merged and report back.

williamrandolph commented:
@dakrone I think I can get back to work on this, but since the PR has been sitting so long, I'm going to close it and put the work out in smaller chunks, linking back to here as necessary.
